baremaps-postgres/src/main/java/com/baremaps/postgres/jdbi/FeatureMapper.java

```diff
@@ -0,0 +1,59 @@
+package com.baremaps.postgres.jdbi;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import org.jdbi.v3.core.mapper.RowMapper;
+import org.jdbi.v3.core.statement.StatementContext;
+
+public class FeatureMapper implements RowMapper<Feature> {
+
+  private final GeometryColumnMapper mapper = new GeometryColumnMapper();
+
+  @Override
+  public Feature map(ResultSet rs, StatementContext ctx) throws SQLException {
+    var feature = new Feature();
+    feature.setProperties(new HashMap<>());
+
+    var metadata = ctx.getStatement().getMetaData();
+    var columnCount = metadata.getColumnCount();
+    for (int i = 1; i <= columnCount; i++) {
+      var columnName = metadata.getColumnName(i);
+      switch (columnName) {
+        case "id":
+        case "fid":
+        case "gid":
+          // Set first occurrence as id
+          if (feature.getId() == null) {
+            feature.setId(rs.getString(i));
+          }
+          // Set following occurrence as properties
+          else {
+            feature.getProperties().put(columnName, rs.getString(i));
+          }
+          break;
+        case "type":
+          feature.setType(rs.getString(i));
+          break;
+        default:
+          var columnTypeName = metadata.getColumnTypeName(i);
+          switch (columnTypeName) {
+            case "geometry":
+            case "geometry(geometry)":
+            case "geometry(point)":
+            case "geometry(linestring)":
+            case "geometry(polygon)":
+              // Set first occurrences as geometry and ignore remaining ones
+              if (feature.getGeometry() == null) {
+                feature.setGeometry(mapper.map(rs, i, ctx));
+              }
+              break;
+            default:
+              feature.getProperties().put(columnName, rs.getString(i));
```

Comment:
So double or int value converted in string representation? or did I miss something ;-)

Reply:
This is naive experiment level stuff ;-)

In my understanding, the feature api should have some support for types, but I havn't been able to find a consise description of the Feature class in the documentation of the OGC. Any pointer?

Reply:
> In my understanding, the feature api should have some support for types, but I havn't been able to find a consise description of the Feature class in the documentation of the OGC. Any pointer?

Nope. sorry.  It's a bit mysterious
It's  the same thing with the concepts of property and geometry.  Should we consider geometry as a property of the feature or as a particular type ?
If we opt for the last solution it means that we may have to add a map to store the geometries for each column.

Reply:
Here is the definition of a feature (featureGeoJSON.yaml) in the OGC API.

```
type: object
required:
  - type
  - geometry
  - properties
properties:
  type:
    type: string
    enum:
      - Feature
  geometry:
    $ref: geometryGeoJSON.yaml
  properties:
    type: object
    nullable: true
  id:
    oneOf:
      - type: string
      - type: integer
  links:
    type: array
    items:
      $ref: link.yaml
```

In fact, properties is just an object, so I guess that on the client side it could have properties of different type (in JSON). For instance:

```json
{ 
  "properties": {
    "string": "string",
    "number": 1.5,
    "boolean": true,
    "object": {},
    "array": []
  }
}
```

The challenge will probably be to map the types in a consistent fashion all the way from json, java, and postgis.

Reply:
A possibility for mapping the properties may be to use arrays which are pretty standard in databases. For instance, something like

```sql
SELECT '1' as id, 'Feature' as type, 'POINT(1 1)' as geometry, ['key1', 'value1', 'key2', 'value2'] as properties;
SELECT '1' as id, 'Feature' as type, 'POINT(1 1)' as geometry, ['key1', 'key2'] as keys, ['value1', 'value2'] as values;
```

Reply:
Another possibility may be to use the json type present in a lot of databases (postgres, mysql, h2) to store properties.

Reply:
For me,  Feature has nothing todo with geojson encoding definition. It's an abstraction to define how properties are modeled for spatial feature.
GeoJSON is limited to one geometry column and for me a Feature can support multiple column geometries.
It is true that the specification is ambiguous on this point. And the link to https://www.w3.org/TR/sdw-bp/#spatial-things-features-and-geometry , does not provide a concrete answer.

Reply:
Ok, I see. I looked at different documentations and geotools as a good model to represent features.
https://docs.geotools.org/latest/userguide/library/opengis/feature.html